src/main/java/io/confluent/connect/hdfs/DataWriter.java

```diff
@@ -380,50 +339,79 @@ public void recover(TopicPartition tp) {
-      topics.add(tp.topic());
-    }
+    // ensure each topic is only synced once
+    Set<String> topics = topicPartitionWriters.keySet()
+        .stream().map(TopicPartition::topic).collect(Collectors.toSet());
+
+    for (String topic : topics) {
+      Path recoveredFileWithMaxOffsets = getLatestFilePathForTopic(topic);
+
+      if (recoveredFileWithMaxOffsets != null) {
+        Schema latestSchema = schemaFileReader
+            .getSchema(connectorConfig, recoveredFileWithMaxOffsets);
 
-    try {
-      for (String topic : topics) {
         String topicDir = FileUtils.topicDirectory(
             connectorConfig.url(),
-            topicDirs.get(topic),
+            connectorConfig.getTopicsDirFromTopic(topic),
             topic
         );
-        CommittedFileFilter filter = new TopicCommittedFileFilter(topic);
-        FileStatus fileStatusWithMaxOffset = FileUtils.fileStatusWithMaxOffset(
-            storage,
-            new Path(topicDir),
-            filter
-        );
-        if (fileStatusWithMaxOffset != null) {
-          final Path path = fileStatusWithMaxOffset.getPath();
-          final Schema latestSchema;
-          latestSchema = schemaFileReader.getSchema(
-              connectorConfig,
-              path
-          );
-          hive.createTable(hiveDatabase, topic, latestSchema, partitioner);
-          List<String> partitions = hiveMetaStore.listPartitions(hiveDatabase, topic, (short) -1);
-          FileStatus[] statuses = FileUtils.getDirectories(storage, new Path(topicDir));
-          for (FileStatus status : statuses) {
-            String location = status.getPath().toString();
-            if (!partitions.contains(location)) {
-              String partitionValue = getPartitionValue(location);
-              hiveMetaStore.addPartition(hiveDatabase, topic, partitionValue);
-            }
-          }
+        attemptCreatingHiveTable(topic, topicDir, latestSchema);
+      }
+    }
+  }
+
+  /**
+   * Attempt to create a Hive table based on a file's schema from the topic.
+   *
+   * @param topic the name of the topic
+   * @param topicDir the directory of the topic
+   * @param latestSchema the schema of the latest file
+   */
+  private void attemptCreatingHiveTable(String topic, String topicDir, Schema latestSchema) {
+    try {
+      hive.createTable(hiveDatabase, topic, latestSchema, partitioner);
+      List<String> partitions = hiveMetaStore.listPartitions(hiveDatabase, topic, (short) -1);
+      FileStatus[] statuses = FileUtils.getDirectories(storage, new Path(topicDir));
+      for (FileStatus status : statuses) {
+        String location = status.getPath().toString();
+        if (!partitions.contains(location)) {
+          String partitionValue = getPartitionValue(location);
+          hiveMetaStore.addPartition(hiveDatabase, topic, partitionValue);
         }
       }
+    } catch (HiveMetaStoreException e) {
+      log.warn("Could not create Hive table: {}", e.getMessage());
```

Comment:
why are we catching this here? shouldn't we fail?

Reply:
@levzem good point, we should fail to remain consistent with previous behavior.